File: Tema 05 - Typescript/PetShelterFrontend/src/services/FundraiserService.ts
import axios from 'axios';
import { Person } from '../models/Person';
import { Fundraiser } from '../models/Fundraiser';

export class FundraiserService
{
    private apiUrl: string = "http://localhost:5009";
    public getAll(): Promise<Fundraiser[]> {
        return axios
        .get(this.apiUrl + '/Fundraiser')
        .then(response => {
            //console.log(response.data);

            var fundraiserResponse: Fundraiser[] = [];
            response.data.forEach((fundraiserFromApi:Fundraiser) => {

                let owner = new Person(fundraiserFromApi.owner.name, fundraiserFromApi.owner.id);
                fundraiserFromApi.owner=owner;
                fundraiserResponse.push(fundraiserFromApi);
            });

            return fundraiserResponse;
        })
    }
    public addFundraiser(fundraiser: Fundraiser): Promise<void> {
        //let owner:PersonDto
        let fundraiserToAdd: FundraiserDto= {
            name: fundraiser.name,
            dueDate :fundraiser.dueDate,
            goalValue:fundraiser.goalValue,
            owner : {
                dateOfBirth : new Date(),
                idNumber : fundraiser.owner.id,
                name : fundraiser.owner.name
            },
        }

        return axios.post(this.apiUrl + '/Fundraiser', fundraiserToAdd);
    }
    public getById(id: number):Promise<Fundraiser>
    {
    return axios.get(this.apiUrl + `/Fundraiser/${id}`)
        .then(response => {
            let fundraiserResponse:Fundraiser=
            Fundraiser.createFundraiserWithAllAttributes(response.data.name, response.data.owner, response.data.goalValue, 
                 response.data.dueDate, response.data.status, response.data.creationDate,response.data.donationAmount, response.data.donors);
            return fundraiserResponse;
    })
    }

    public deleteFundraiser(id: number):Promise<void>
    {
        return axios.put(this.apiUrl + `/Fundraiser/${id}/delete`);
    }

    public donate(fundraiserId: number, donorId: number, amount:number):Promise<void>
    {
        let donation:any ={amount:amount}
        return axios.put(this.apiUrl + `/Fundraiser/${fundraiserId}/donate/${donorId}`, donation);
    }
}

interface PersonDto
{
    name: string; 
    dateOfBirth: Date | undefined; 
    idNumber: string; 
}

interface FundraiserDto
{
    name:string;
    owner: PersonDto;
    goalValue: number;
    dueDate:Date;
}

File: Tema 03 - Web API/PetShelter/PetShelter.Domain/Services/FundraiserService.cs
using PetShelter.DataAccessLayer.Models;
using PetShelter.DataAccessLayer.Repository;
using PetShelter.Domain.Exceptions;
using PetShelter.Domain.Extensions.DataAccess;
using PetShelter.Domain.Extensions.DomainModel;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PetShelter.Domain.Services
{
    public class FundraiserService : IFundraiserService
    {
        private readonly IFundraiserRepository _fundraiserRepository;
        private readonly IPersonRepository _personRepository;
        public FundraiserService(IFundraiserRepository fundraiserRepository, IPersonRepository personRepository)
        {
            _fundraiserRepository = fundraiserRepository;
            _personRepository = personRepository;
        }

        public async Task<int> CreateFundraiserAsync(Person owner, Fundraiser fundraiser)
        {
            var person = await _personRepository.GetOrAddPersonAsync(owner.FromDomainModel());
            var addedFundraiser = new DataAccessLayer.Models.Fundraiser
            {
                Name = fundraiser.Name,
                Owner=person,
                OwnerId=person.Id,
                GoalValue=fundraiser.GoalValue,
                DueDate=fundraiser.DueDate,
                Status=fundraiser.Status.ToString(),
                CreationDate=fundraiser.CreationDate,
                DonationAmount=fundraiser.DonationAmount,
                Donors=fundraiser.Donors.Select(x => x.FromDomainModel()).ToList()
            };
            await _fundraiserRepository.Add(addedFundraiser);
            return addedFundraiser.Id;
    }

        public async Task DeleteFundraiserAsync(int fundraiserId)
        {
            var fundraiser = await _fundraiserRepository.GetById(fundraiserId);
            if(fundraiser==null)
            {
                throw new NotFoundException($"Fundraiser with id {fundraiserId} not found.");
            }
            fundraiser.Status = "Closed";
            _fundraiserRepository.Update(fundraiser);
        }

        public async Task DonateToFundraiserAsync(int fundraiserId, int donorId, decimal amount)
        {
            var fundraiser = await _fundraiserRepository.GetById(fundraiserId);
            if(fundraiser==null)
            {
                throw new NotFoundException($"Fundraiser with id {fundraiserId} not found.");
            }
            var person = await _personRepository.GetById(donorId);
            if(person==null)
            {
                throw new NotFoundException($"Person with id {donorId} not found.");
            }
            if(fundraiser.Status=="Closed")
            {
                throw new NotFoundException($"Fundraiser with id {fundraiserId} is closed.");
            }
            person.FundraiserWhichGotTheDonationId = fundraiserId;
            fundraiser.DonationAmount += amount;
            if(fundraiser.DonationAmount>fundraiser.GoalValue)
            {
                fundraiser.Status = "Closed";
            }
            _fundraiserRepository.Update(fundraiser);
            _personRepository.Update(person);
        }

        public async Task<IReadOnlyCollection<Fundraiser>> GetAllFundraisers()
        {
            var fundraisers = await _fundraiserRepository.GetAll();
            return fundraisers.Select(async p => { p.Owner = await _personRepository.GetById(p.OwnerId);
                p.Donors = await _fundraiserRepository.GetDonors(p.Id); return p; })
                .Select(p => p.Result)
                .Select(p => p.ToDomainModel())
                .ToImmutableArray();
        }

        public async Task<Fundraiser> GetFundraiser(int fundraiserId)
        {
            var fundraiser = await _fundraiserRepository.GetById(fundraiserId);
            if (fundraiser == null)
            {
                return null;
            }
            fundraiser.Owner = await _personRepository.GetById(fundraiser.OwnerId);
            fundraiser.Donors = await _fundraiserRepository.GetDonors(fundraiserId);
            return fundraiser.ToDomainModel(); ;
        }
    }
}
